src/Datadog.Trace.Build/DatadogLogger.cs

```diff
@@ -0,0 +1,275 @@
+            eventSource.BuildFinished += EventSource_BuildFinished;
+            eventSource.ProjectStarted += EventSource_ProjectStarted;
+            eventSource.ProjectFinished += EventSource_ProjectFinished;
+            eventSource.ErrorRaised += EventSource_ErrorRaised;
+            eventSource.WarningRaised += EventSource_WarningRaised;
+        }
+
+        /// <inheritdoc />
+        public void Shutdown()
+        {
+        }
+
+        private void EventSource_BuildStarted(object sender, BuildStartedEventArgs e)
+        {
+            Log.Debug("Build Started");
+
+            _buildSpan = _tracer.StartSpan(BuildTags.BuildOperationName);
+            _buildSpan.SetMetric(Tags.Analytics, 1.0d);
+            _buildSpan.SetTraceSamplingPriority(SamplingPriority.UserKeep);
+
+            _buildSpan.Type = SpanTypes.Build;
+            _buildSpan.SetTag(BuildTags.BuildName, e.SenderName);
+            foreach (KeyValuePair<string, string> envValue in e.BuildEnvironment)
+            {
+                _buildSpan.SetTag($"{BuildTags.BuildEnvironment}.{envValue.Key}", envValue.Value);
+            }
+
+            _buildSpan.SetTag(BuildTags.BuildCommand, Environment.CommandLine);
+            _buildSpan.SetTag(BuildTags.BuildWorkingFolder, Environment.CurrentDirectory);
+            _buildSpan.SetTag(BuildTags.BuildStartMessage, e.Message);
+
+            _buildSpan.SetTag(BuildTags.RuntimeOSArchitecture, Environment.Is64BitOperatingSystem ? "x64" : "x86");
+            _buildSpan.SetTag(BuildTags.RuntimeProcessArchitecture, Environment.Is64BitProcess ? "x64" : "x86");
+
+            CIEnvironmentValues.DecorateSpan(_buildSpan);
+        }
+
+        private void EventSource_BuildFinished(object sender, BuildFinishedEventArgs e)
+        {
+            Log.Debug("Build Finished");
+
+            _buildSpan.SetTag(BuildTags.BuildStatus, e.Succeeded ? BuildTags.BuildSucceededStatus : BuildTags.BuildFailedStatus);
+            if (!e.Succeeded)
+            {
+                _buildSpan.Error = true;
+            }
+
+            _buildSpan.SetTag(BuildTags.BuildEndMessage, e.Message);
+            _buildSpan.Finish(e.Timestamp);
+        }
+
+        private void EventSource_ProjectStarted(object sender, ProjectStartedEventArgs e)
+        {
+            if (e.TargetNames.StartsWith("_"))
+            {
+                // Ignoring internal targetNames
+                return;
+            }
+
+            Log.Debug("Project Started");
+
+            int parentContext = e.ParentProjectBuildEventContext.ProjectContextId;
+            int context = e.BuildEventContext.ProjectContextId;
+            if (!_projects.TryGetValue(parentContext, out Span parentSpan))
+            {
+                parentSpan = _buildSpan;
+            }
+
+            string projectName = Path.GetFileName(e.ProjectFile);
+
+            Span projectSpan = _tracer.StartSpan(BuildTags.BuildOperationName, parent: parentSpan.Context, serviceName: projectName);
```

Comment:
Does this handle `BuildInParallel`? I imagine this is why you start the span with a specific parent context in mind so that parallel builds can all have the correct parent

Reply:
Yeah, that's right, msbuild gives me the parent context with `e.ParentProjectBuildEventContext.ProjectContextId` as an int, that's why I store the span `parent/child` relationship with a `ConcurrentDictionary<int, Span>`